core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -132,6 +135,24 @@ final void setStream(ClientStream stream) {
     drainPendingCalls();
   }
 
+  protected boolean isStreamTransferCompleted() {
+    return realStreamStarted.getCount() == 0;
+  }
+
+  protected void awaitStreamTransferCompletion() {
+    // Wait until accepted RPCs transfer to the real stream so that we can properly cancel or
+    // shutdown. Not waiting for transfer completion may cause pending calls orphaned. #636.
+    boolean delegationComplete;
+    try {
+      delegationComplete = realStreamStarted.await(5, TimeUnit.SECONDS);
+    } catch (InterruptedException ex) {
+      delegationComplete = false;
+    }
+    if (!delegationComplete) {
+      Thread.currentThread().interrupt();
```

Comment:
To clarify this method is called at user cancel, shutdownNow() cancel, and reprocess once we detect shutdown() has been called. The interruption is handled no where, it will be propagated to the user(at ManagedChannelImpl main thread).

This path is only executed inside SyncContext where we only handle uncaught exceptions and put the channel into panic. And it looks like it is the only limited exception handling mechanism here at client side.

I have been struggling when the failure handling here and am open for discussion. So I thought that either we have to enhance our exception propagation path, e.g. distinguish different exceptions and handle them at higher granularity. Before that, keeping it silent here looks generally inappropriate (today it is almost like that), so I interrupted the caller thread. The interruption can be handled at SyncContext, but then it seems better to propagate it out as well.

Reply:
> The interruption is handled no where, it will be propagated to the user(at ManagedChannelImpl main thread).

If the calling thread doesn't check its interrupted bit or is blocked on `wait/join/sleep` etc, it never knows. I don't think there is any case that this happens.

I was expecting such an await to be somehow hooked up with Channel's `awaitTermination()`. That is, instead of doing synchronous wait, we use Future. Have you tried with that approach?

Reply:
> I was expecting such an await to be somehow hooked up with Channel's awaitTermination(). That is, instead of doing synchronous wait, we use Future. Have you tried with that approach?

I think it's already like that: awaitTermination() will be waiting for awaitTransferCompletion() here. So once awaitTransferCompletion() is done, dealyedTransport will be reporting termination through listener to ManagedChannelImpl to do further shutdown routine. Here just we set a time bound to awaitTransferCompletion(), as I thought that we perhaps should not wait forever(e.g. customer never call start() so the transfer would never be done. theoretically this can happen, but it won't for now, as we control that in ClientCallImpl shortly after we create the stream). But waiting forever is also an option.

Reply:
> I think it's already like that: awaitTermination() will be waiting for awaitTransferCompletion() here.

Now shutdownNow() may be synchronously waiting for awaitTransferCompletion() to complete. I don't think we want synchronous waits in the shutdown path. I am thinking of it'd be a callback for transfer completed (or cancelled), instead of a latch that needs to be waited on.

Reply:
Hm, potentially we can have settableFuture installed into delayedStream during reprocess() and then listen on the event when real stream started, and at the listener callback we check the shutdown status.